test/natvis_tests.cpp

```diff
@@ -0,0 +1,185 @@
+    fca_set_begin, fca_set_end, fca_multiset_begin, fca_multiset_end);
+  use(foa_flat_map_begin, foa_flat_map_end, foa_flat_set_begin,
+    foa_flat_set_end, foa_node_map_begin, foa_node_map_end, foa_node_set_begin,
+    foa_node_set_end);
+
+  int break_here = 0;
+  use(break_here);
+}
+
+class offset_ptr_tester_
+{
+  static constexpr std::size_t SEGMENT_SIZE = 64 * 1024;
+  std::string segment_name = to_string(boost::uuids::random_generator()());
+  boost::interprocess::managed_shared_memory segment{
+    boost::interprocess::create_only, segment_name.c_str(), SEGMENT_SIZE};
+
+  using map_allocator = boost::interprocess::allocator<map_value_type,
+    boost::interprocess::managed_shared_memory::segment_manager>;
+  using set_allocator = boost::interprocess::allocator<set_value_type,
+    boost::interprocess::managed_shared_memory::segment_manager>;
+
+  template <template <class...> class MapTemplate>
+  using map_type = MapTemplate<std::string, int, boost::hash<std::string>,
+    std::equal_to<std::string>, map_allocator>;
+  template <template <class...> class SetTemplate>
+  using set_type = SetTemplate<std::string, boost::hash<std::string>,
+    std::equal_to<std::string>, set_allocator>;
+
+public:
+  offset_ptr_tester_()
+  {
+    boost::interprocess::shared_memory_object::remove(segment_name.c_str());
+  }
+  ~offset_ptr_tester_()
+  {
+    boost::interprocess::shared_memory_object::remove(segment_name.c_str());
+  }
+
+  template <template <class...> class MapTemplate>
+  map_type<MapTemplate>* construct_map()
+  {
+    return &*segment.construct<map_type<MapTemplate> >(
+      typeid(map_type<MapTemplate>).name())(
+      map_allocator(segment.get_segment_manager()));
+  }
+  template <template <class...> class SetTemplate>
+  set_type<SetTemplate>* construct_set()
+  {
+    return &*segment.construct<set_type<SetTemplate> >(
+      typeid(set_type<SetTemplate>).name())(
+      set_allocator(segment.get_segment_manager()));
+  }
+} offset_ptr_tester;
+
+class default_tester_
+{
+  template <template <class...> class MapTemplate>
+  using map_type = MapTemplate<std::string, int, boost::hash<std::string>,
+    std::equal_to<std::string>, std::allocator<map_value_type> >;
+  template <template <class...> class SetTemplate>
+  using set_type = SetTemplate<std::string, boost::hash<std::string>,
+    std::equal_to<std::string>, std::allocator<set_value_type> >;
+
+public:
+  template <template <class...> class MapTemplate>
+  std::unique_ptr<map_type<MapTemplate> > construct_map()
+  {
+    return std::make_unique<map_type<MapTemplate> >();
+  }
+  template <template <class...> class SetTemplate>
+  std::unique_ptr<set_type<SetTemplate> > construct_set()
+  {
+    return std::make_unique<set_type<SetTemplate> >();
+  }
+} default_tester;
+
+#endif // defined(BOOST_MSVC)
+
+int main()
+{
+#if defined(BOOST_MSVC)
+
+  natvis_test(default_tester);
+  natvis_test(offset_ptr_tester);
+
+#endif // defined(BOOST_MSVC)
```

Comment:
One thing to consider might be refactoring this to have only one `#ifdef BOOST_MSVC` and two alternate mains.

The shorter main can live at the top and have something like a `BOOST_PRAGMA_MESSAGE("skipping test for unsupported platform")`